src/Google.Logging.Log4Net/MemoryLogQueue.cs

```diff
@@ -0,0 +1,119 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Google.Logging.Log4Net
+{
+    internal class MemoryLogQueue : ILogQueue
+    {
+        private struct LogEntryExtraSize
+        {
+            public LogEntryExtraSize(LogEntryExtra logEntryExtra, int size)
+            {
+                LogEntryExtra = logEntryExtra;
+                Size = size;
+            }
+
+            public LogEntryExtra LogEntryExtra { get; }
+            public int Size { get; }
+        }
+
+        public MemoryLogQueue(long maxMemorySize, int maxMemoryCount)
+        {
+            _maxMemorySize = maxMemorySize;
+            _maxMemoryCount = maxMemoryCount;
+        }
+
+        private readonly object _lockObj = new object();
+        private readonly long _maxMemorySize;
+        private readonly int _maxMemoryCount;
+
+        private readonly Queue<LogEntryExtraSize> _q = new Queue<LogEntryExtraSize>();
+        private long _currentMemorySize = 0;
+        private long _maxIdSeen = -1;
+
+        public DateTimeRange EnqueueAsync(IEnumerable<LogEntryExtra> logEntries)
+        {
+            lock (_lockObj)
+            {
+                DateTimeRange lostRange = null;
+                foreach (var entry in logEntries)
+                {
+                    if ((_maxMemoryCount > 0 && _q.Count >= _maxMemoryCount) || (_maxMemorySize > 0 && _currentMemorySize > _maxMemorySize))
+                    {
+                        var lostEntry = _q.Dequeue();
+                        _currentMemorySize -= lostEntry.Size;
+                        if (lostRange == null)
+                        {
+                            lostRange = new DateTimeRange(lostEntry.LogEntryExtra.Entry.Timestamp(), lostEntry.LogEntryExtra.Entry.Timestamp());
+                        }
+                        else
+                        {
+                            lostRange = lostRange.WithTo(lostEntry.LogEntryExtra.Entry.Timestamp());
+                        }
+                    }
+                    var sizedEntry = new LogEntryExtraSize(entry, _maxMemorySize > 0 ? entry.Entry.CalculateSize() : 0);
+                    _q.Enqueue(sizedEntry);
+                    if (entry.Id > _maxIdSeen)
+                    {
+                        _maxIdSeen = entry.Id;
+                    }
+                    _currentMemorySize += sizedEntry.Size;
+                }
+                return lostRange;
+            }
+        }
+
+        public Task<long> GetNextIdAsync()
+        {
+            lock (_lockObj)
```

Comment:
Everything else is consistently locked with _lockObj. I would like to keep that consistency.
This method is only called once, so there's no performance issue ;)